cobalt/watchdog/watchdog.cc

```diff
@@ -497,11 +550,71 @@ bool Watchdog::Unregister(const std::string& name, bool lock) {
   return result;
 }
 
+std::shared_ptr<Client> Watchdog::UnregisterByClient(
+    std::shared_ptr<Client> client) {
+  if (is_disabled_) return nullptr;
+
+  starboard::ScopedLock scoped_lock(mutex_);
+
+  std::string name = "";
+  if (client) name = client->name;
+
+  // Unregisters.
+  for (auto it = client_list_.begin(); it != client_list_.end(); it++) {
+    if (client == *it) {
+      client_list_.erase(it);
+      SB_DLOG(INFO) << "[Watchdog] Unregistered: " << name;
+      return nullptr;
+    }
+  }
+  return client;
+}
+
 bool Watchdog::Ping(const std::string& name) { return Ping(name, ""); }
 
 bool Watchdog::Ping(const std::string& name, const std::string& info) {
   if (is_disabled_) return true;
 
+  starboard::ScopedLock scoped_lock(mutex_);
+
+  auto it = client_map_.find(name);
+  bool client_exists = it != client_map_.end();
+
+  if (client_exists) {
+    Client* client = it->second.get();
+    return PingHelper(client, name, info);
+  }
+  SB_DLOG(ERROR) << "[Watchdog] Unable to Ping: " << name;
+  return false;
+}
+
+std::shared_ptr<Client> Watchdog::PingByClient(std::shared_ptr<Client> client) {
+  return PingByClient(client, "");
+}
+
+std::shared_ptr<Client> Watchdog::PingByClient(std::shared_ptr<Client> client,
+                                               const std::string& info) {
+  if (is_disabled_) return client;
+
+  std::string name = "";
+  if (client) name = client->name;
+
+  starboard::ScopedLock scoped_lock(mutex_);
+
+  for (auto it = client_list_.begin(); it != client_list_.end(); it++) {
+    if (client == *it) {
+      if (PingHelper(client.get(), name, info)) {
+        return client;
+      }
+      return nullptr;
+    }
+  }
+  SB_DLOG(ERROR) << "[Watchdog] Unable to Ping: " << name;
+  return nullptr;
```

Comment:
ditto. Is it enough to just return a boolean?

Reply:
ack